src/dist/replication/meta_server/meta_split_service.cpp

```diff
@@ -131,5 +131,146 @@ void meta_split_service::do_app_partition_split(std::shared_ptr<app_state> app,
+        return;
+    }
+
+    if (parent_context.stage == config_status::pending_proposal ||
+        parent_context.stage == config_status::pending_remote_sync) {
+        dwarn_f("another request is syncing with remote storage, ignore this request");
+        return;
+    }
+
+    ddebug_f("parent({}) will resgiter child({})", parent_gpid, child_gpid);
+    parent_context.stage = config_status::pending_remote_sync;
+    parent_context.msg = rpc.dsn_request();
+    parent_context.pending_sync_task = add_child_on_remote_storage(rpc, true);
+}
+
+dsn::task_ptr meta_split_service::add_child_on_remote_storage(register_child_rpc rpc,
+                                                              bool create_new)
+{
+    const auto &request = rpc.request();
+    const std::string &partition_path = _state->get_partition_path(request.child_config.pid);
+    blob value = dsn::json::json_forwarder<partition_configuration>::encode(request.child_config);
+    if (create_new) {
+        return _meta_svc->get_remote_storage()->create_node(
+            partition_path,
+            LPC_META_STATE_HIGH,
+            std::bind(&meta_split_service::on_add_child_on_remote_storage_reply,
+                      this,
+                      std::placeholders::_1,
+                      rpc,
+                      create_new),
+            value);
+    } else {
+        return _meta_svc->get_remote_storage()->set_data(
+            partition_path,
+            value,
+            LPC_META_STATE_HIGH,
+            std::bind(&meta_split_service::on_add_child_on_remote_storage_reply,
+                      this,
+                      std::placeholders::_1,
+                      rpc,
+                      create_new),
+            _meta_svc->tracker());
+    }
+}
+
+void meta_split_service::on_add_child_on_remote_storage_reply(error_code ec,
+                                                              register_child_rpc rpc,
+                                                              bool create_new)
+{
+    zauto_write_lock(app_lock());
+
+    const auto &request = rpc.request();
+    auto &response = rpc.response();
+
+    std::shared_ptr<app_state> app = _state->get_app(request.app.app_id);
+    dassert_f(app != nullptr, "app is not existed, id({})", request.app.app_id);
+    dassert_f(app->status == app_status::AS_AVAILABLE || app->status == app_status::AS_DROPPING,
+              "app is not available now, id({})",
+              request.app.app_id);
+
+    dsn::gpid parent_gpid = request.parent_config.pid;
+    dsn::gpid child_gpid = request.child_config.pid;
+    config_context &parent_context = app->helpers->contexts[parent_gpid.get_partition_index()];
+
+    if (ec == ERR_TIMEOUT ||
+        (ec == ERR_NODE_ALREADY_EXIST && create_new)) { // retry register child on remote storage
+        bool retry_create_new = (ec == ERR_TIMEOUT) ? create_new : false;
+        int delay = (ec == ERR_TIMEOUT) ? 1 : 0;
+        parent_context.pending_sync_task =
```

Comment:
Yes, `pending_sync_task` should be set every time when updating `partition_configuration` on zk, it will guarantee configuration won't be updated concurrently. You can see codes like `config_context.pending_sync_task = update_configuration_on_remote(req)` in `server_state.cpp`.